packages/frontend/core/src/components/affine/auth/oauth.tsx

```diff
@@ -64,9 +64,22 @@ function OAuthProvider({
   popupWindow: (url: string) => void;
 }) {
   const serverService = useService(ServerService);
+  const auth = useService(AuthService);
   const { icon } = OAuthProviderMap[provider];
 
   const onClick = useCallback(() => {
+    if (scheme && scheme !== 'web') {
+      auth
+        .oauthPreflight(provider, scheme)
+        .then(oauthUrl => {
+          popupWindow(oauthUrl);
```

Comment:
The most critical change is that the preflight is initiated by the client, not the browser.